### PR TITLE
RFC: improve inlined line numbers, take 2

### PR DESCRIPTION
**Pull request by JeffBezanson**
This is a rebased and edited version of #14949. To reduce the scope a bit, I changed it to store file names and function names instead of references to LambdaInfos. The goal is to get something mergeable as soon as possible, as this change has several important fixes from @carnaval :
1. It fixes the formatting of our LLVM debug info and represents inlined functions properly.
2. It changes the representation of source location info in our IR.
3. It adds the ability to represent any number of nested inlined functions in backtraces.

I want to get these changes in even if the backtraces are not perfect yet. We can improve it later, either by adding back the LambdaInfo (or Method) references or by keeping a mapping from filename and line number to Methods.

I will merge #15583 into this branch, so we can take advantage of some of this for macro code locations as well.

### CODE REVIEW

File: src/debuginfo.cpp [hunk 1]
```diff
@@ -759,6 +769,10 @@ extern "C" void jl_register_fptrs(uint64_t sysimage_base, void **fptrs, jl_lambd
     sysimg_fvars_n = n;
 }
 
+#ifdef _OS_LINUX_
+#include <link.h>
```
Comment (tkelman) [inline src/debuginfo.cpp]: top of file, around L53-54?

**Comment by JeffBezanson** [comment]
I'm not sure how to handle locations from macro expansions. My guess is that there is not good support for the filename magically switching in the middle of a function, so for now I handled it the same way as an inlined function (since that seems to work well), with the inlined function name being `macro expansion`.

**Comment by toivoh** [comment]
Will there be a way for Julia code examining the AST to figure or which file that lines from a macro expansion came from?

**Comment by StefanKarpinski** [comment]
> I'm not sure how to handle locations from macro expansions.

I think ultimately the only way to handle generated code sanely will be to keep generated ASTs and be able to step through them. The `julia` program should have a command-line argument `--keep-source=[yes|no]` which defaults to "yes" in interactive mode and defaults to "no" in non-interactive mode.

**Comment by JeffBezanson** [comment]
> Will there be a way for Julia code examining the AST to figure or which file that lines from a macro expansion came from?

Yes, there are `push_loc` and `pop_loc` for blocks from a different file.

**Comment by timholy** [comment]
Big fan of this effort.

> there are push_loc and pop_loc for blocks from a different file.

Awesome! Doesn't yet seem documented in this PR, though. Might need a small section in devdocs or something.

File: src/codegen.cpp [hunk 2]
```diff
@@ -4286,7 +4277,7 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
         #ifndef LLVM34
         SP = dbuilder.createFunction((DIDescriptor)dbuilder.getCU(),
         #else
-        SP = dbuilder.createFunction(CU,
+        SP = dbuilder.createFunction(topfile,
```
Comment (Keno) [inline src/codegen.cpp]: I'm not sure this works. If the function isn't associated to a compile unit, it might not end up being represented in the binary.

File: src/codegen.cpp [hunk 3]
```diff
@@ -4675,84 +4668,64 @@ static std::unique_ptr<Module> emit_function(jl_lambda_info_t *lam, jl_llvm_func
...
+                if (jl_array_len(stmt_e->args) > 2)
+                    inl_name = jl_symbol_name((jl_sym_t*)jl_exprarg(stmt_e, 2));
+                else
+                    inl_name = "macro expansion";
+                SP = dbuilder.createFunction(new_file,
+                                             inl_name + ";",
+                                             inl_name,
+                                             new_file,
+                                             0,
+                                             jl_di_func_sig,
```
Comment (Keno) [inline src/codegen.cpp]: This isn't really correct. NOt that anything currently depends on it, but we might want to revisit this at a later point.

File: src/debuginfo.cpp [hunk 4]
```diff
@@ -814,9 +825,23 @@ bool jl_dylib_DI_for_fptr(size_t pointer, const llvm::object::ObjectFile **obj,
         jl_in_stackwalk = 0;
 #else // ifdef _OS_WINDOWS_
     Dl_info dlinfo;
-    if ((dladdr((void*)pointer, &dlinfo) != 0) && dlinfo.dli_fname) {
+    int dladdr_success;
+    uint64_t fbase;
+#ifdef _OS_LINUX_
+    // dlinfo.dli_fbase is not the right value for the main executable on linux
```
Comment (Keno) [inline src/debuginfo.cpp]: Should this comment go below?

File: src/debuginfo.cpp [hunk 5]
```diff
@@ -1228,34 +1249,83 @@ void jl_getFunctionInfo(char **name, char **filename, size_t *line,
...
+};
+enum section_type {
+    Sect_Code, Sect_Data, Sect_RoData
+};
+static section_memory *section_mem[3];
+#include <sys/mman.h>
+#ifdef _OS_DARWIN_
+#define MAP_ANONYMOUS MAP_ANON
+#endif
+uint8_t *juliaAllocateSection(uintptr_t size, unsigned align, section_type typ)
```
Comment (Keno) [inline src/debuginfo.cpp]: This shouldn't be part of this PR.

**Comment by JeffBezanson** [comment]
Dev docs added and comments addressed. Should be ready to merge.

File: test/backtrace.jl [hunk 6]
```diff
@@ -4,68 +4,15 @@ bt = backtrace()
...
+    if lkup[1][1] == :backtrace
+        @test lkup[1][5] == false # fromC
         have_backtrace = true
         break
     end
 end
 
 @test have_backtrace
 
-# Test location information for inlined code (ref issues #1334 #12544)
```
Comment (tkelman) [inline test/backtrace.jl]: if the results of these all changed, can we change them to test the affirmative of what they should be now, instead of deleting them altogether?

Comment (JeffBezanson) [inline test/backtrace.jl]: Ah yes, will do.

File: src/task.c [hunk 7]
```diff
@@ -617,13 +621,19 @@ void jl_init_tasks(void)
...
+                                            jl_symbol("bufsz"),
+                                           jl_symbol("stkbuf"),
+                                           jl_symbol("ssize")),
+                                   jl_svec(13,
+                                           jl_any_type,
+                                           jl_any_type, jl_sym_type,
+                                           jl_any_type, jl_any_type,
+                                           jl_any_type, jl_any_type,
+                                           jl_any_type, jl_any_type,
+                                           jl_tupletype_fill(sizeof(jl_jmp_buf), jl_uint8_type),
```
Comment (tkelman) [inline src/task.c]: needs a cast to `(jl_value_t *) jl_uint8_type`

**Comment by tkelman** [comment]
This breaks building against LLVM 3.3, if we're still concerned with trying to keep that working for testing/perf comparison.
https://gist.github.com/02d8c53f6c9e5962cb738292478646bb

**Comment by carnaval** [comment]
Thanks a lot. I got sidetracked with the memory issues and was starting to lose faith ;). Glad to see a lot of it was salvageable. Allocating our own jit sections was an easy win if we don't care too much about W^X but I'm not sure that'll be the consensus.

**Comment by JeffBezanson** [comment]
Yeah the allocation part is definitely important to work on, but should just be done separately. I suspect it's probably ok to use RWX pages by default, with a switch to disable them.

**Comment by JeffBezanson** [comment]
It would make me happy to remove all the LLVM version ifdefs. If we have ifdefs for some LLVM version, but it's not tested in CI and even most developers never run it, is it *really* supported? The readability and maintainability gains would be significant.

However currently I'm having trouble building with LLVM 3.3. I get errors about `std::unique_ptr` and such. There must be another switch I need to flip.

**Comment by tkelman** [comment]
`CXXFLAGS=-std=c++11`

I agree that it would help clean things up, but 3.7 is still a pretty bad compiler perf regression relative to 3.3 isn't it? Though I see a failure at the end of bootstrap that was introduced by one of the codegen rewrites in March, so maybe it's already broken and not salvagable.

**Comment by yuyichao** [comment]
You need `CXXFLAGS+=' -std=c++0x'` to compile with llvm 3.3 with gcc. I've also seen issue about linker confused by '@' in the symbol table though.

**Comment by JeffBezanson** [comment]
Thanks.

> 3.7 is still a pretty bad compiler perf regression relative to 3.3 isn't it?

Yes it is. But we already decided to make that tradeoff.

**Comment by tkelman** [comment]
On master, as a default build configuration. We don't have every-commit CI for lots of nominally supported configurations, ARM, FreeBSD, etc. We haven't deleted all of the LLVM 3.3 code yet, so the decision of whether or not to keep that a possible option for testing hasn't totally been finalized.

**Comment by yuyichao** [comment]
I believe @vtjnash is testing it on Mac, which doesn't require much special configuration due to clang being permissive about c++11 features and `@` not being a special character for the linker. (FWIW g++ 6.1 use gnu++14 by default so the `CXXFLAGS` setting shouldn't be necessary anymore)

**Comment by JeffBezanson** [comment]
OK, LLVM 3.3 build restored to at least as working as it was before.

**Comment by vtjnash** [comment]
Is this the cause of the new failure in genstdlib?
```
WARNING: missing docs for signature:

    lookup(pointer::Union{Ptr{Void}, UInt}) -> StackFrame
```

**Comment by tkelman** [comment]
81d2c464f0f8e1390e61de0a9ffc87f8b435314b
